### PR TITLE
Add probability param for random flip functions.

### PR DESCRIPTION
**Pull request by pabloduque0**
Based on the conversation in #37 . Will work on adding one or two functions in the upcoming weeks.

Apologies for the extra unnecessary commits, vs studio butcher the notebook and had to revert that one, we can squash it. Happy to update the notebook in a follow up as well.

A few thoughts:

- I have not added checks (asserts) on the prob value to allow jitting without needing static args. Let me know if this is okay, Im not super familiar with chex so maybe Im missing something there.
- I had to change the way the extra parameters are evaluated in the test to kwargs always. I think this should be fine as long as we dont make some of those parameter positional only. Let me know if you rather take a different approach on this one.
- I love how clever the test setup is! 🚀🚀🚀

Feel free to add any comments on the wording of the docstring as well.

### CODE REVIEW

File: dm_pix/_src/augment.py [hunk 1]
```diff
@@ -295,16 +295,60 @@ def solarize(image: chex.Array, threshold: chex.Numeric) -> chex.Array:
...
+      flip_left_right,
+      lambda x: x,
+      image)
 
 
-def random_flip_up_down(key: chex.PRNGKey, image: chex.Array) -> chex.Array:
-  """50% chance of `flip_up_down(...)` otherwise returns image unchanged."""
-  coin_flip = jax.random.bernoulli(key)
-  return jax.lax.cond(coin_flip, flip_up_down, lambda x: x, image)
+def random_flip_up_down(
```
Comment (claudiofantacci) [inline dm_pix/_src/augment.py:326]: Can you please apply the same comments for `random_flip_left_right` here 😄 ?

**Review by claudiofantacci — COMMENTED** [review]
Thanks, great PR! 🚀 

Can you please either:
 - remove the commits where you first and and then remove code
 - squash the commits into a single one

File: dm_pix/_src/augment.py [hunk 2]
```diff
@@ -295,16 +295,60 @@ def solarize(image: chex.Array, threshold: chex.Numeric) -> chex.Array:
...
+    probability: chex.Numeric = .5
+) -> chex.Array:
+  """Applies `flip_left_right(...)` or not with a given probability.
+
+  If the number drawn from a uniform [0, 1] falls bellow the given probability
+  it applies `flip_up_down(...)` to the image, otherwise returns original image.
+
+  Args:
+    key: key for pseudo-random number generator.
+    image: a JAX array which represents an image.
```
[the remark this answers is not in the document]
Comment (claudiofantacci) [reply dm_pix/_src/augment.py:311]: You're right indeed! Feel free to update the docstring (or revert to yours) wherever that is applicable in this file 😄

**Comment by pabloduque0** [comment]
Thanks a lot for the review! Let me know if I missed any of the comments

File: dm_pix/_src/augment.py [hunk 3]
```diff
@@ -295,16 +295,62 @@ def solarize(image: chex.Array, threshold: chex.Numeric) -> chex.Array:
...
 
-def random_flip_left_right(key: chex.PRNGKey, image: chex.Array) -> chex.Array:
-  """50% chance of `flip_left_right(...)` otherwise returns image unchanged."""
-  coin_flip = jax.random.bernoulli(key)
-  return jax.lax.cond(coin_flip, flip_left_right, lambda x: x, image)
+def random_flip_left_right(
+    key: chex.PRNGKey,
+    image: chex.Array,
+    *,
+    probability: chex.Numeric = .5,
```
Comment (claudiofantacci) [inline dm_pix/_src/augment.py:302]: Sorry to be pedantic 😅
`.5` -> `0.5`

File: dm_pix/_src/augment.py [hunk 4]
```diff
@@ -295,16 +295,62 @@ def solarize(image: chex.Array, threshold: chex.Numeric) -> chex.Array:
...
+      flip_left_right,
+      lambda x: x,
+      image)
+
+
+def random_flip_up_down(
+    key: chex.PRNGKey,
+    image: chex.Array,
+    *,
+    probability: chex.Numeric = .5,
```
Comment (claudiofantacci) [inline dm_pix/_src/augment.py:331]: Same here, sorry to be pedantic 😅
`.5` -> `0.5`

**Review by claudiofantacci — CHANGES_REQUESTED** [review]

File: dm_pix/_src/augment.py (hunk 3 above)
Comment (pabloduque0) [reply dm_pix/_src/augment.py]: Oh no, its perfect! Good point!

File: dm_pix/_src/augment.py (hunk 4 above)
Comment (pabloduque0) [reply dm_pix/_src/augment.py]: ✅

**Comment by pabloduque0** [comment]
Let me know if there is anything left to iron out :)

File: dm_pix/_src/augment.py [hunk 5]
```diff
@@ -160,7 +160,7 @@ def flip_left_right(
   Assumes that the image is either ...HWC or ...CHW and flips the W axis.
 
   Args:
-    image: an RGB image, given as a float tensor in [0, 1].
+    image: a JAX array which represents an image.
```
Comment (claudiofantacci) [inline dm_pix/_src/augment.py:163]: Use the same wordings 😄 

```
image: a JAX array representing an image. Assumes that the image is either ...HWC or ...CHW.
```

File: dm_pix/_src/augment.py [hunk 6]
```diff
@@ -183,7 +183,7 @@ def flip_up_down(
   Assumes that the image is either ...HWC or ...CHW, and flips the H axis.
 
   Args:
-    image: an RGB image, given as a float tensor in [0, 1].
+    image: a JAX array which represents an image.
```
Comment (claudiofantacci) [inline dm_pix/_src/augment.py:186]: Use the same wordings 😄 
```
image: a JAX array representing an image. Assumes that the image is either ...HWC or ...CHW.
```

File: dm_pix/_src/augment.py [hunk 7]
```diff
@@ -295,16 +295,62 @@ def solarize(image: chex.Array, threshold: chex.Numeric) -> chex.Array:
...
+    probability: chex.Numeric = 0.5,
+) -> chex.Array:
+  """Applies `flip_left_right(...)` with a given probability.
+
+  If the number drawn from a uniform [0, 1] falls bellow the given probability
+  it applies `flip_up_down(...)` to the image, otherwise returns original image.
+
+  Args:
+    key: a JAX RNG key.
+    image: a JAX array which represents an image. Assumes that the image is
```
Comment (claudiofantacci) [inline dm_pix/_src/augment.py:311]: `a JAX array which represents an image` -> `a JAX array representing an image`

File: dm_pix/_src/augment.py [hunk 8]
```diff
@@ -295,16 +295,62 @@ def solarize(image: chex.Array, threshold: chex.Numeric) -> chex.Array:
...
+    probability: chex.Numeric = 0.5,
+) -> chex.Array:
+  """Applies `flip_up_down(...)` with a given probability.
+
+  If the number drawn from a uniform [0, 1] falls bellow the given probability
+  it applies `flip_up_down(...)` to the image, otherwise returns original image.
 
+  Args:
+    key: a JAX RNG key.
+    image: a JAX array which represents an image. Assumes that the image is
```
Comment (claudiofantacci) [inline dm_pix/_src/augment.py:340]: `a JAX array which represents an image` -> `a JAX array representing an image`

**Review by claudiofantacci — CHANGES_REQUESTED** [review]
Final comments! 🚀

File: dm_pix/_src/augment.py (hunk 5 above)
Comment (pabloduque0) [reply dm_pix/_src/augment.py]: Good point! fixed in all 4

**Review by claudiofantacci — APPROVED** [review]
Thanks a lot @pabloduque0! 🚀

**Comment by claudiofantacci** [comment]
Once internal review is passed, the PR will be automatically merged!
It might take a bit of time, but rest assure it will go through 😄 
Thanks again for contributing, and feel free to open any other PR! 🚀

**Comment by claudiofantacci** [comment]
@pabloduque0, we have a release freeze in effect for NeurIPS 2022 until end of 2022/05/19.

After that date, the PR will be automatically merged!
Thanks for you patience 😄

**Comment by pabloduque0** [comment]
@claudiofantacci no worries at all! No rush from my side, after all I have the changes locally to use them :)

Yes, I will look into adding a couple functions in the near future, might follow up on #37 to check with you if they are a good fit or not.

Thanks a lot again for the review and the merge!

**Pull request merged**